packages-exp/functions-exp/src/error.ts

```diff
@@ -0,0 +1,176 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { HttpsError, FunctionsErrorCode } from '@firebase/functions-types-exp';
+import { Serializer } from './serializer';
+import { HttpResponseBody } from './service';
+
+/**
+ * Standard error codes for different ways a request can fail, as defined by:
+ * https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
+ *
+ * This map is used primarily to convert from a backend error code string to
+ * a client SDK error code string, and make sure it's in the supported set.
+ */
+const errorCodeMap: { [name: string]: FunctionsErrorCode } = {
+  OK: 'ok',
+  CANCELLED: 'cancelled',
+  UNKNOWN: 'unknown',
+  INVALID_ARGUMENT: 'invalid-argument',
+  DEADLINE_EXCEEDED: 'deadline-exceeded',
+  NOT_FOUND: 'not-found',
+  ALREADY_EXISTS: 'already-exists',
+  PERMISSION_DENIED: 'permission-denied',
+  UNAUTHENTICATED: 'unauthenticated',
+  RESOURCE_EXHAUSTED: 'resource-exhausted',
+  FAILED_PRECONDITION: 'failed-precondition',
+  ABORTED: 'aborted',
+  OUT_OF_RANGE: 'out-of-range',
+  UNIMPLEMENTED: 'unimplemented',
+  INTERNAL: 'internal',
+  UNAVAILABLE: 'unavailable',
+  DATA_LOSS: 'data-loss'
+};
+
+/**
+ * An explicit error that can be thrown from a handler to send an error to the
+ * client that called the function.
+ */
+export class HttpsErrorImpl extends Error implements HttpsError {
```

Comment:
Should it implement `FirebaseError` instead?

Reply:
So the tricky part is that `HttpsError` has a field called "details" which is type unknown because it is whatever arbitrary data the developer has decided to attach to the error thrown from the cloud function. Functions SDK tries to clean it up a bit but the end result still could be anything from null to a primitive to a keyed object. The details are passed here: https://github.com/firebase/firebase-js-sdk/blob/1177fa6924d562ad3d70c0972f61625e9737ff7d/packages-exp/functions-exp/src/error.ts#L175

`FirebaseError` extends Error plus `ErrorData` which only allows StringLike | undefined values. If "details" was just one type of thing I would maybe just expand ErrorData's value options but since it can be pretty much anything I don't want to ruin ErrorData's specificity by adding "unknown".

Any thoughts?

Reply:
We can make `HttpsErrorImpl` extend `FirebaseError` instead of `Error`, so it will have `details` in addition to all the fields that `FirebaseError` has.

We should probably rename the class to `FirebaseFunctionsError` and export it, so people can use it as a type or also `instanceof` to handle it differently.

Reply:
Maybe we should extend `ErrorData` to allow any values, is there any reason its current specificity is important?

Regardless, we should still define a subclass error for functions, because it will provide stronger typing than generic string index.

Reply:
Done, I think. I don't know the purpose of `ErrorData`'s specificity. Something to do with the strict `ErrorFactory` setup I assume.

I went with `FunctionsError` because the naming pattern we came up with in the API review was to not put "firebase" in front of "functions" so this seems consistent.